newrelic/hooks/external_pyzeebe.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2010 New Relic, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from newrelic.api.application import application_instance
+from newrelic.common.object_wrapper import wrap_function_wrapper
+from newrelic.api.transaction import current_transaction, add_custom_attribute
+from newrelic.api.time_trace import add_custom_span_attribute
+from newrelic.api.background_task import BackgroundTask
+from newrelic.api.function_trace import FunctionTrace
+
+# Adds client method params as txn or span attributes
+def _add_client_input_attributes(method_name, add_attr, args, kwargs):
+    if method_name in ("run_process", "run_process_with_result"):
+        bpmn_id = kwargs.get("bpmn_process_id", args[0] if args else None)
+        if bpmn_id is not None:
+            add_attr("zeebe.client.bpmnProcessId", bpmn_id)
+    elif method_name == "publish_message":
+        msg_name = kwargs.get("name", args[0] if args else None)
+        if msg_name is not None:
+            add_attr("zeebe.client.messageName", msg_name)
+        correlation_key = kwargs.get("correlation_key", args[1] if args and len(args) > 1 else None)
+        if correlation_key is not None:
+            add_attr("zeebe.client.correlationKey", correlation_key)
+        message_id = kwargs.get("message_id")
+        if message_id is None and len(args) > 4:
+            message_id = args[4]
+        if message_id is not None:
+            add_attr("zeebe.client.messageId", message_id)
+    elif method_name == "deploy_resource":
+        resources = list(args)
+        if len(resources) == 1 and isinstance(resources[0], (list, tuple)):
+            resources = list(resources[0])
+        if resources:
+            add_attr("zeebe.client.resourceCount", len(resources))
+            if len(resources) == 1:
+                try:
+                    add_attr("zeebe.client.resourceFile", str(resources[0]))
+                except Exception:
+                    add_attr("zeebe.client.resourceFile", repr(resources[0]))
+
+
+# Async wrapper that instruments a ZeebeClient method.
+def _nr_client_wrapper(method_name):
+    async def wrapper(wrapped, instance, args, kwargs):
+        txn = current_transaction()
+        if txn:
+            add_fn = add_custom_span_attribute
```

Comment:
To add the new pyzeebe attributes, you may want to check out the `add_agent_attribute` function for this rather than adding these as custom attributes. Here are some examples of its use in other places within the agent: http://github.com/newrelic/newrelic-python-agent/blob/main/newrelic/hooks/messagebroker_kombu.py#L204 https://github.com/newrelic/newrelic-python-agent/blob/main/newrelic/hooks/messagebroker_kafkapython.py#L145. You will also want to add all of these new pyzeebe attributes into this set: https://github.com/newrelic/newrelic-python-agent/blob/main/newrelic/core/attribute.py#L43

Reply:
@umaannamalai - I tried using _add_agent_attribute_ and it yields undesired behavior when adding those attributes in the existing `txn` if block. It will add them to the parent span instead of each of the individual `FunctionTrace` spans (i.e - `run_process`, `deploy_resource`, etc). What is your recommendation to achieve stamping those attributes on each individual span? Below is an example screenshot where those client methods are called inside a `main()` function that is annotated with `@newrelic.agent.background_task()`

<img width="1538" alt="image" src="https://github.com/user-attachments/assets/43cddac0-206a-427d-ba8e-1d52e05df8b9" />

Reply:
@umaannamalai - following up here